File: ChordVoicingsCreator/src/main/java/com/campoy/chord/voicing/creator/model/tinwhistle/SevenHoleWhistle6Front1Flat6LowerBackThumb.java
package com.campoy.chord.voicing.creator.model.tinwhistle;

import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

import com.campoy.chord.voicing.creator.model.musictheory.Interval;
import com.campoy.chord.voicing.creator.model.musictheory.Note;
import com.campoy.chord.voicing.creator.model.musictheory.Key;
import com.google.common.collect.Sets;

import lombok.Data;

@Data
public class SevenHoleWhistle6Front1Flat6LowerBackThumb {

    private final Note baseNote;
    
    private Set<NoteAndFingering> naturalNotes = null;
    private Set<NoteAndFingering> crossFingeredNotes = null;
    private Set<NoteAndFingering> allPossibleNotes = null;
    
    public Set<NoteAndFingering> getNaturalNotes(){
        if(naturalNotes == null) {
            Set<NoteAndFingering> notes = new HashSet<>();
            for( WhistleFingering crossFingering : WhistleFingerings.naturalFingeringsWithLowerBackThumbHole ) {
                notes.add( new NoteAndFingering(baseNote.up(crossFingering.getSemitonesFromRoot()), 
                                                crossFingering));
            }
            naturalNotes = notes;
        }
        return naturalNotes;
    }
    
    public Set<NoteAndFingering> getCrossFingeredNotes(){
        if(crossFingeredNotes == null) {
            Set<NoteAndFingering> notes = new HashSet<>();
            for( WhistleFingering crossFingering : WhistleFingerings.crossFingeringsWithLowerBackFlat6ThumbHole ) {
                notes.add( new NoteAndFingering(baseNote.up(crossFingering.getSemitonesFromRoot()), 
                                                crossFingering));
            }
            crossFingeredNotes = notes;
        }
        return crossFingeredNotes;
    }
    

    
    public Set<NoteAndFingering> getAllPossibleNotesWithFingerings(){
        if(allPossibleNotes == null) {
            allPossibleNotes = Sets.union(getNaturalNotes() , getCrossFingeredNotes());
        }
        return allPossibleNotes;
    }
    
    public Set<Note> getAllPossibleNotes(){
        return getAllPossibleNotesWithFingerings()
               .stream()
               .map(noteAndFingering -> noteAndFingering.getNote())
               .collect(Collectors.toSet());
    }
    
    public List<NoteAndFingering> getNeededCrossFingerings(List<Note> scaleNotes){
        return getCrossFingeredNotes()
                .stream()
                .filter(noteAndFingering -> scaleNotes.contains(noteAndFingering.getNote()))
                .collect(Collectors.toList());
    }
    
    public List<NoteAndFingering> getFingerings(List<Note> scaleNotes){
        return getAllPossibleNotesWithFingerings()
                .stream()
                .filter(noteAndFingering -> scaleNotes.contains(noteAndFingering.getNote()))
                .collect(Collectors.toList());
    }
}

File: src/model/tinwhistle/SevenHoleWhistle6Front1Flat6LowerBackThumb.ts
import { EqualsTrait } from "../general/EqualsTrait.js";
import { Note } from "../musictheory/Note.js";
import { WhistleFingering } from "./WhistleFingering.js";
import { WhistleFingerings } from "./WhistleFingerings.js";

export class SevenHoleWhistle6Front1Flat6LowerBackThumb implements EqualsTrait{
    private baseNote: Note;
    private naturalNotes: Set<NoteAndFingering> | null = null;
    private crossFingeredNotes: Set<NoteAndFingering> | null = null;
    private allPossibleNotes: Set<NoteAndFingering> | null = null;
  
    constructor(baseNote: Note) {
      this.baseNote = baseNote;
    }
    equals(other: any): boolean {
      if(!(other instanceof SevenHoleWhistle6Front1Flat6LowerBackThumb)){
        return false;
      }
      return this.baseNote.equals(other.baseNote);
    }
  
    public getNaturalNotes(): Set<NoteAndFingering> {
      if (this.naturalNotes === null) {
        const notes = new Set<NoteAndFingering>();
        for (const naturalFingering of WhistleFingerings.naturalFingeringsWithLowerBackThumbHole) {
          notes.add(new NoteAndFingering(this.baseNote.up(naturalFingering.getSemitonesFromRoot()), naturalFingering));
        }
        this.naturalNotes = notes;
      }
      return this.naturalNotes;
    }
  
    public getCrossFingeredNotes(): Set<NoteAndFingering> {
      if (this.crossFingeredNotes === null) {
        const notes = new Set<NoteAndFingering>();
        for (const crossFingering of WhistleFingerings.crossFingeringsWithLowerBackFlat6ThumbHole) {
          notes.add(new NoteAndFingering(this.baseNote.up(crossFingering.getSemitonesFromRoot()), crossFingering));
        }
        this.crossFingeredNotes = notes;
      }
      return this.crossFingeredNotes;
    }
  
    public getAllPossibleNotesWithFingerings(): Set<NoteAndFingering> {
      if (this.allPossibleNotes === null) {
        this.allPossibleNotes = new Set([...this.getNaturalNotes(), ...this.getCrossFingeredNotes()]);
      }
      return this.allPossibleNotes;
    }
  
    public getAllPossibleNotes(): Set<Note> {
      return new Set([...this.getAllPossibleNotesWithFingerings()].map(noteAndFingering => noteAndFingering.getNote()));
    }
  
    public getNeededCrossFingerings(scaleNotes: Note[]): NoteAndFingering[] {
      return [...this.getCrossFingeredNotes()].filter(noteAndFingering => scaleNotes.includes(noteAndFingering.getNote()));
    }
  
    public getFingerings(scaleNotes: Note[]): NoteAndFingering[] {
      return [...this.getAllPossibleNotesWithFingerings()].filter(noteAndFingering => scaleNotes.includes(noteAndFingering.getNote()));
    }
  }
  
  class NoteAndFingering {
    private note: Note;
    private fingering: WhistleFingering;
  
    constructor(note: Note, fingering: WhistleFingering) {
      this.note = note;
      this.fingering = fingering;
    }
  
    public getNote(): Note {
      return this.note;
    }
  
    public getFingering(): WhistleFingering {
      return this.fingering;
    }
  }